### PR TITLE
Remove duplicate setState in TextFormFieldState.reset

### LINKED ISSUE #68997: [proposal] avoid TextFormField.reset() to call setState() twice
I was reviewing the `TextFormField` code and I found its `reset` code calls `setState`:

https://github.com/flutter/flutter/blob/bced0351aedd719f65429d02b0baa8c8653342ea/packages/flutter/lib/src/material/text_form_field.dart#L346-L352

But before that, `reset` calls `super.reset()`, and this's the code for `super.reset()`:

https://github.com/flutter/flutter/blob/bced0351aedd719f65429d02b0baa8c8653342ea/packages/flutter/lib/src/widgets/form.dart#L442-L450

So in the end there will be 2 calls to `setState()` from inside the `reset()` function.

Not a major bug or performance impact, but a minor improvement.

### PR DESCRIPTION
**Pull request by justinmc**
## Description

TextFormFieldState.reset was calling setState twice.  This removes one call.

## Related Issues

Closes https://github.com/flutter/flutter/issues/68997

## Tests

None.  It seems impractical to test the number of times setState is called.  If someone disagrees and knows a robust way to do so, let me know and I'll add a test.

### CODE REVIEW

**Comment by justinmc** [comment]
@LongCatIsLooong @feinstein Any other ideas  on how to test something like this?  The test that I wrote here passes, but it doesn't catch the problem if I remove the fix.  I would expect that without this fix, there would be 3 builds, but there are still only 2.  Maybe we are batching setState calls under the hood or something like that?

**Comment by feinstein** [comment]
AFAIK `setState` will mark the widget as dirty, so calling it multiple times wouldn't trigger more builds, just set a variable that's already set.

The minor performance impact that I mentioned is only about running the `setState` function code itself.

**Comment by feinstein** [comment]
If you really want to test this, maybe you would have to subclass the `TextFormField` and `override` `setState`, adding a counter there? I just wonder if a test like this is useful to be kept in the framework for ever, maybe to avoid regressions? Don't know, just thinking out loud.

File: packages/flutter/test/material/text_form_field_test.dart [hunk 1]
```diff
@@ -406,6 +406,35 @@ void main() {
     expect(find.text('initialValue'), findsOneWidget);
   });
 
+  // Regression test for https://github.com/flutter/flutter/issues/68997.
+  testWidgets('reset calls setState only once', (WidgetTester tester) async {
+    int builds = 0;
```
Comment (LongCatIsLooong) [inline packages/flutter/test/material/text_form_field_test.dart:411]: I thought this shouldn't fail even with the additional `setState`? The validator should only be run once for each frame (in `build`), no?

**Comment by LongCatIsLooong** [comment]
Oh sorry didn't see the comments. I think it would be a little weird to write a test to enforce it just here: it's a pretty common practice throughout the codebase. The performance impact would be almost negligible IMO, as the `setState` call shortcircuits when the element is already marked as dirty. Maybe try to get a test exemption?

**Comment by feinstein** [comment]
agreed. Also the order of the `super` calling shouldn't make a difference, as `setState` will be called anyaway. but I think it's nice as it shows the intention.

**Comment by justinmc** [comment]
Thanks for the advice on testing, I have removed my test.

**Comment by Hixie** [comment]
There's nothing wrong with calling setState multiple times (it doesn't trigger multiple builds).

What we _should_ test is that the various codepaths that call reset() all end up triggering a build. I don't see where setState is being called in this file, maybe it's in the superclass? We should at the very minimum add a comment in the places where we're changing state but not calling setState to say _where_ the setState gets invoked.

**Comment by feinstein** [comment]
@Hixie yes, setState is being called on the superclass. 

What I meant with this issue wasn't about triggering build multiple times, it was just a tiny improvement for not calling the function itself, running the asserts on debug and etc. Nothing major like avoiding builds.

I like the idea to document whenever setState is being called by a superclass or other non-visible mechanism.

**Comment by justinmc** [comment]
I've added the comment, good idea.

It seems that `reset` is only called publicly, and that is [already tested](https://github.com/flutter/flutter/blob/7ceb20417c93553f4cbd7a5a42bb2878968f6775/packages/flutter/test/material/text_form_field_test.dart#L387) in a way that confirms that a rebuild happened.  If anyone thinks I should add any other specific test, let me know.  Otherwise ready for re-review again.

**Comment by Hixie** [comment]
LGTM

**Pull request merged**